### PR TITLE
Rename AMD GPU architectures

### PR DESCRIPTION
**Pull request by Rombur**
This is a proposal to rename the AMD architectures in Kokkos. Currently to enable an AMD GPU we need to use names like `Kokkos_ARCH_VEGA90A`, i.e. `VEGA` + the GFX flag or `NAVI` + the GFX flag. This works pretty well for the NAVI cards but it's terrible for the VEGA cards. The code name Vega corresponds to the MI50/60 cards.  The code name for MI100 is Arcturus and the code name for the MI200 series is Aldebaran. Every year the Vega name becomes more and more obsolete, so we need a new name. At first, I was thinking about using MI50, MI100, etc. The problem is that an MI50 and an MI60 are identical from our point of view. There is a similar issue with the MI200 series. CMake also has support for AMD GPU and they simply use `GFX90A`. We could mimic what they are doing and use `Kokkos_ARCH_GFX90A`. I don't find it obvious that we are compiling for an AMD gpu. Instead, inspired by what we are doing for Intel, I propose that we use `Kokkos_ARCH_AMD_GFX90A`. This makes it clear that we are using an AMD GPU and the GFX flag matches what we are giving to the compiler. I also propose to drop our VEGA and NAVI macros that were rarely used and instead introduce an AMD_GPU macro similar to what we do for Intel.  Users can keep using the current names but new GPU will only support the new naming conventions. 

This PR is a draft until we agree on the new naming convention. 

Still missing in the PR: setting `Kokkos_ARCH_AMD_GFX90A` should define both `KOKKOS_ARCH_AMD_GFX90A` and `KOKKOS_ARCH_VEGA90A` and vice versa.

cc: @arghdos

### CODE REVIEW

**Review by arghdos — APPROVED** [review]
I agree with the change, and I think the arch name works well (and is more descriptive than what we currently have)

File: Makefile.kokkos [hunk 1]
```diff
@@ -13,7 +13,7 @@ KOKKOS_DEVICES ?= "Threads"
 # NVIDIA:   Kepler,Kepler30,Kepler32,Kepler35,Kepler37,Maxwell,Maxwell50,Maxwell52,Maxwell53,Pascal60,Pascal61,Volta70,Volta72,Turing75,Ampere80,Ampere86,Ada89,Hopper90
 # ARM:      ARMv80,ARMv81,ARMv8-ThunderX,ARMv8-TX2,A64FX
 # IBM:      BGQ,Power7,Power8,Power9
-# AMD-GPUS: Vega906,Vega908,Vega90A,Navi1030
+# AMD-GPUS: Gfx906,Gfx908,Gfx90A,Gfx1030
```
Comment (arghdos) [inline Makefile.kokkos:16]: Any reason for these to not be fully capitalized? I guess the rest in the list here aren't, but AFAICT this is the only case where they're not.

Comment (Rombur) [reply Makefile.kokkos:16]: You are right, I'll change this.

**Comment by ldh4** [comment]
Will this change break any existing build scripts if a user had manually specified a kokkos arch (i.e. `-DKokkos_ARCH_VEGA906`) in their cmake scripts?

**Comment by Rombur** [comment]
No this will be backward compatible once I fix the export of the macros.

**Comment by nmm0** [comment]
Per the meeting, this will need the docs updated. @dalg24 suggested to put a `*` next to the old names, and have code names/easily searchable values

File: Makefile.kokkos [hunk 2]
```diff
@@ -1075,29 +1075,29 @@ endif
...
   KOKKOS_INTERNAL_HIP_ARCH_FLAG := --offload-arch=gfx1030
 endif
-ifeq ($(KOKKOS_INTERNAL_USE_ARCH_NAVI1100), 1)
-  tmp := $(call kokkos_append_header,"$H""define KOKKOS_ARCH_NAVI1100")
-  tmp := $(call kokkos_append_header,"$H""define KOKKOS_ARCH_NAVI")
+ifeq ($(KOKKOS_INTERNAL_USE_ARCH_AMD_GFX1100), 1)
+  tmp := $(call kokkos_append_header,"$H""define KOKKOS_ARCH_AMD_GFX1100")
+  tmp := $(call kokkos_append_header,"$H""define KOKKOS_ARCH_AMD_GPU")
   KOKKOS_INTERNAL_HIP_ARCH_FLAG := --offload-arch=gfx1100
 endif
```
Comment (masterleinad) [inline Makefile.kokkos:1102]: This means we don't set the old macros anymore when using Makefiles but we do for `CMake`? Shouldn't we be consistent?

Comment (Rombur) [reply Makefile.kokkos:1102]: One more reason to switch to cmake

File: cmake/KokkosCore_config.h.in
Comment (dalg24) [inline cmake/KokkosCore_config.h.in:1]: Maybe add `// deprecated` on the old macros

**Review by dalg24 — COMMENTED** [review]
Do you plan any replacement for the VEGA and NAVI macros (regroup family of similar architectures)

File: cmake/KokkosCore_config.h.in [hunk 3]
```diff
@@ -110,6 +110,12 @@
 #cmakedefine KOKKOS_ARCH_AMD_ZEN
 #cmakedefine KOKKOS_ARCH_AMD_ZEN2
 #cmakedefine KOKKOS_ARCH_AMD_ZEN3
+#cmakedefine KOKKOS_ARCH_AMD_GFX906
+#cmakedefine KOKKOS_ARCH_AMD_GFX908
+#cmakedefine KOKKOS_ARCH_AMD_GFX90A
+#cmakedefine KOKKOS_ARCH_AMD_GFX1030
+#cmakedefine KOKKOS_ARCH_AMD_GFX1100
+#cmakedefine KOKKOS_ARCH_AMD_GPU
```
[the remark this answers is not in the document]
Comment (crtrott) [reply cmake/KokkosCore_config.h.in:118]: not sure that there is a good consistent value to be set for AMD which actually implies growing capability?

**Comment by Rombur** [comment]
> Do you plan any replacement for the VEGA and NAVI macros ?

No I don't. Right now, the difference between VEGA and NAVI is that VEGA is using a wave front of 64 and NAVI use 32. However on some NAVI architectures, you can switch the wavefront to use 64. The current division VEGA/NAVI does not make sense and I  don't feel confident introducing a new 32/64 wavefront macro, since we don't have any NAVI hardware to test the different configurations.

File: cmake/KokkosCore_config.h.in (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rombur) [reply cmake/KokkosCore_config.h.in:118]: > Please confirm you don't want it to be KOKKOS_IMPL_ARCH_AMD_GPU and if so explain why in a few words.

My idea was to have a macro for user to do sanity check or to use in their code if they want to do something special on AMD GPU.

> not sure that there is a good consistent value to be set for AMD which actually implies growing capability?

No, I don't think so.

[the remark this answers is not in the document]
Comment (arghdos) [reply cmake/KokkosCore_config.h.in:118]: The gfxarch ID does at least indicate a card's capabilities as according to the compiler: https://llvm.org/docs/AMDGPUUsage.html#processors, though I agree that it won't necessarily align to 'growing capabilities'

File: Makefile.kokkos
[the remark this answers is not in the document]
Comment (stanmoore1) [reply Makefile.kokkos:1]: Thanks for the note. Maintaining backwards compatibility for Makefiles would be nice, or at least put it into a deprecation cycle.

[the remark this answers is not in the document]
Comment (dalg24) [reply Makefile.kokkos:1]: > Thanks for the note. Maintaining backwards compatibility for Makefiles would be nice, or at least put it into a deprecation cycle.

The HIP old architecture names or the whole makefiles shenanigans?

File: Makefile.kokkos [hunk 4]
```diff
@@ -402,10 +402,10 @@ ifeq ($(KOKKOS_INTERNAL_USE_ARCH_ZEN3), 0)
     KOKKOS_INTERNAL_USE_ARCH_ZEN := $(call kokkos_has_string,$(KOKKOS_ARCH),Zen)
   endif
 endif
-KOKKOS_INTERNAL_USE_ARCH_VEGA906 := $(call kokkos_has_string,$(KOKKOS_ARCH),Vega906)
-KOKKOS_INTERNAL_USE_ARCH_VEGA908 := $(call kokkos_has_string,$(KOKKOS_ARCH),Vega908)
-KOKKOS_INTERNAL_USE_ARCH_VEGA90A := $(call kokkos_has_string,$(KOKKOS_ARCH),Vega90A)
-KOKKOS_INTERNAL_USE_ARCH_NAVI1030 := $(call kokkos_has_string,$(KOKKOS_ARCH),Navi1030)
+KOKKOS_INTERNAL_USE_ARCH_AMD_GFX906 := $(or $(call kokkos_has_string,$(KOKKOS_ARCH),VEGA906),$(call kokkos_has_string,$(KOKKOS_ARCH),AMD_GFX906))
```
Comment (dalg24) [inline Makefile.kokkos:405]: Did you mean to change the case?  Vega906 -> VEGA906

File: Makefile.kokkos
[the remark this answers is not in the document]
Comment (dalg24) [reply Makefile.kokkos:1]: Clarification, what changes is not the configuration keyword, these we will continue to honor.
In the current form of the PR we would not define the `KOKKOS_ARCH_VEGA*` macros which Bruno actually looked for in LAMMPS and they are not used.

Sorry about the confusion.

File: Makefile.kokkos (hunk 4 above)
Comment (dalg24) [reply Makefile.kokkos:405]: This needs fixing though.

Comment (Rombur) [reply Makefile.kokkos:405]: I can change it back but it makes no difference because `kokkos_has_string`  converts the lower case to upper case.

**Review by dalg24 — APPROVED** [review]

**Comment by Rombur** [comment]
There is a problem with the architecture auto-detection. I am working on fixing the issue.

**Comment by Rombur** [comment]
It should be good now. I just had to move function to a different place. I also updated the OpenACC and OpenMPTarget backends to use the new macros.

**Comment by Rombur** [comment]
Can we get this merged? I think @arghdos is waiting on this to be merged before creating a PR for MI300

**Review by masterleinad — APPROVED** [review]

**Pull request merged**